bigtable/client/table_apply_test.cc

```diff
@@ -92,3 +92,29 @@ TEST(TableApplyTest, Retry) {
   EXPECT_NO_THROW(table.Apply(bigtable::SingleRowMutation(
       "bar", {bigtable::SetCell("fam", "col", 0, "val")})));
 }
+
+/// @test Verify that Table::Apply() retries only idempotent mutations.
+TEST(TableApplyTest, RetryIdempotent) {
+  MockClient client;
+  using namespace ::testing;
+  EXPECT_CALL(client, Stub())
+      .WillRepeatedly(
+          Invoke([&client]() -> google::bigtable::v2::Bigtable::StubInterface& {
+            return *client.mock_stub;
+          }));
+  EXPECT_CALL(*client.mock_stub, MutateRow(_, _, _))
+      .WillOnce(
+          Return(grpc::Status(grpc::StatusCode::UNAVAILABLE, "try-again")));
+
+  bigtable::Table table(&client, "foo_table");
+
+  try {
+    table.Apply(bigtable::SingleRowMutation(
+        "not-idempotent", {bigtable::SetCell("fam", "col", "val")}));
+  } catch (bigtable::PermanentMutationFailures const& ex) {
+    ASSERT_EQ(ex.failures().size(), 1UL);
+    EXPECT_EQ(ex.failures()[0].original_index(), 0);
+  } catch (...) {
+    ADD_FAILURE();
```

Comment:
Can this output the exception as a string so that it's visible what failed and why rather than nothing?

Reply:
I captured the most common exception type and printed it out for that case.  In general, there is no portable way to print out exceptions of unknown type.  You can raise anything in C++, including `void*`, or `MyClassThatIsNotPrintable`.  In C++11 you can at least get the current exception as a "thing":

```C++
} catch(...) {
  std::exception_ptr ex = std::current_exception();
  ...
}
```

But there is very little you can do with that exception other than raise it and pass it around for somebody else to raise.